# Table of contents:
## 4.5 Data Consistensy (theory)
#### Testing mixed data types
#### Finding missing values
#### Finding duplicates
## Task
#### Checking for mixed-type data
#### Missing Values
#### Duplicated Values
#### Exporting the cleaned orders data set

# 4.5 Data Consistency (theory)

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Import datasets
path = r'C:\Users\raque\Documents\08-2023 Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

### Testing mixed data types

In [4]:
# Create a dataframe
df_test = pd.DataFrame()

In [5]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [7]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [8]:
# Changing the data type for df_test
df_test['mix'] = df_test['mix'].astype('str')

### Finding missing values

In [9]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
# Creating a subset to show missing values in product_name
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
df_prods.shape

(49693, 5)

In [13]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [14]:
df_prods_clean.shape

(49677, 5)

### Finding duplicates

In [15]:
# Storing the duplicated values in df_dups
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

In [20]:
# Exporting df_prods_clean_no_dups
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task

In [21]:
df_ords.describe().astype(int)

Unnamed: 0  order_id  user_id  order_number  order_day_of_the_week  \
count     3421083   3421083  3421083       3421083                3421083   
mean      1710541   1710542   102978            17                      2   
std        987581    987581    59533            17                      2   
min             0         1        1             1                      0   
25%        855270    855271    51394             5                      1   
50%       1710541   1710542   102689            11                      3   
75%       2565811   2565812   154385            23                      5   
max       3421082   3421083   206209           100                      6   

       order_hour_of_day  days_since_last_order  
count            3421083                3214874  
mean                  13                     11  
std                    4                      9  
min                    0                      0  
25%                   10                      4  
50%                   13                      7  
75%                   16                     15  
max                   23                     30

Days of the week and Hour of the day start at 0. Meaning it goes from 0 to 6 days or 23 hours. 
The maximum value of days_since_last order is 3, which is a bit odd since the 50% percentile is 7. 
It happens something similar with order number, where the maximum value is 1 while the 25th percentile is 5. 

### Checking for mixed-type data

In [22]:
# Check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There isn't mixed-typed data

### Missing Values

In [23]:
# Looking for missing values/null
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
order_day_of_the_week         0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [24]:
# Creating a subset to show missing values in days_since_last_order
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [25]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_day_of_the_week  \
0                 0   2539329        1             1                      2   
11               11   2168274        2             1                      2   
26               26   1374495        3             1                      1   
39               39   3343014        4             1                      6   
45               45   2717275        5             1                      3   
...             ...       ...      ...           ...                    ...   
3420930     3420930    969311   206205             1                      4   
3420934     3420934   3189322   206206             1                      3   
3421002     3421002   2166133   206207             1                      6   
3421019     3421019   2227043   206208             1                      1   
3421069     3421069   3154581   206209             1                      3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

days_since_last_order is the only column with missing values, with a total of 206209, which is a 6% of the data. 
A possible explanation for this is, if a customer has 1 transaction only, there would no be any values stored in this column. 
This gets confirmed if we have a look at the table above, where the order_number in the missing values is 1. 
Because there's actually nothing wrong with this missing values, we don't need to address them. 

### Duplicated values

In [26]:
# Storing the duplicated values in df_ords_dups
df_ords_dups = df_ords[df_ords.duplicated()]

In [27]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_the_week, order_hour_of_day, days_since_last_order]
Index: []

There aren't any duplicate values

### Exporting the cleaned orders data set

In [28]:
# Exporting df_prods_clean_no_dups
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))